### PR TITLE
[Collections] Cache GitHub API responses

### PR DESCRIPTION
**Pull request by yim-lee**
Motivation:
The `getMetadata` API makes GitHub API requests to fetch additional package metadata. We can cache the results as optimization, which also decreases the chance of caller hitting API limit.

Modifications:
Add cache to GitHubPackageMetadataProvider. Default TTL is 1 hour.

### CODE REVIEW

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]

File: Sources/PackageCollections/Providers/GitHubPackageMetadataProvider.swift [hunk 1]
```diff
@@ -42,6 +45,12 @@ struct GitHubPackageMetadataProvider: PackageMetadataProvider {
             return callback(.failure(Errors.invalidGitURL(reference.location)))
         }
 
+        if let cachedMetadata = self.cache?[reference] {
+            if cachedMetadata.timestamp + DispatchTimeInterval.seconds(self.configuration.cacheTTLInSeconds) > DispatchTime.now() {
```
Comment (tomerd) [inline Sources/PackageCollections/Providers/GitHubPackageMetadataProvider.swift:49]: should we delete the entry if expired?

Comment (yim-lee) [reply Sources/PackageCollections/Providers/GitHubPackageMetadataProvider.swift:49]: We should, though if an entry has expired it also means it will be updated (unless there is failure).

I guess I am more concerned with expired entries that are never fetched again for some reason. Maybe we should limit the number of items stored in the cache and delete expired items whenever we reach the limit?

**Comment by yim-lee** [comment]
@swift-ci please smoke test

File: Sources/PackageCollections/Providers/GitHubPackageMetadataProvider.swift (hunk 1 above)
Comment (yim-lee) [reply Sources/PackageCollections/Providers/GitHubPackageMetadataProvider.swift:52]: We'll delete cache entries when we reach a limit: https://github.com/apple/swift-package-manager/pull/3368/commits/978019722b8749b0f1fbd30c4b04f9e4ff03f25d

File: Sources/PackageCollections/Storage/SQLitePackageCollectionsStorage.swift [hunk 2]
```diff
@@ -829,15 +829,15 @@ final class SQLitePackageCollectionsStorage: PackageCollectionsStorage, Closable
```
Comment (yim-lee) [inline Sources/PackageCollections/Storage/SQLitePackageCollectionsStorage.swift:829]: Unrelated formatting changes

**Comment by yim-lee** [comment]
@swift-ci please smoke test

File: Sources/PackageCollections/Providers/GitHubPackageMetadataProvider.swift [hunk 3]
```diff
@@ -27,11 +27,17 @@ struct GitHubPackageMetadataProvider: PackageMetadataProvider {
     private let diagnosticsEngine: DiagnosticsEngine?
     private let decoder: JSONDecoder
 
+    private let cache: ThreadSafeKeyValueStore<PackageReference, (package: Model.PackageBasicMetadata, timestamp: DispatchTime)>?
+
+    // For cache cleanup
+    private let queue = DispatchQueue(label: "org.swift.swiftpm.GitHubPackageMetadataProvider")
```
Comment (tomerd) [inline Sources/PackageCollections/Providers/GitHubPackageMetadataProvider.swift:33]: not needed?

**Review by tomerd — APPROVED** [review]

**Comment by yim-lee** [comment]
@swift-ci please smoke test
